### PR TITLE
CF-xarray latitude handling

### LINKED ISSUE #1179: non-optional optional argument "lat" for xc.indices.cool_night_index(tasmin: )
### Setup Information

- Xclim version:   0.37.0
- Python version:  3.9.12
- Operating System:  ubuntu linux 20.04.5, 5.14.0-1051-oem, x86_64

### Description

The input array "lat" is considered optional in the documentation, yet if that input array is not given then xc.indices.cool_night_index drops an error:

```TypeError: cool_night_index() missing 1 required positional argument: 'lat'```

Also the source code does not indicate that lat should be considered optional as there is no fallback setting that becomes active if lat is not given.

This may be either a problem of the implementation or of the documentation.

### Steps To Reproduce

_No response_

### Additional context

Making the input of lat optional via changing
```tasmin: xarray.DataArray, lat: xarray.DataArray, freq: str = "YS"```
to
```tasmin: xarray.DataArray, lat: xarray.DataArray | None = None, freq: str = "YS"```
should be programmatically straightforward I believe, but I do not know whether that is scientifically sensible in this case.

### Contribution

- [X] I would be willing/able to open a Pull Request to address this bug.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by Zeitsperre**
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #1179 
- [x] Tests for the changes have been added (for bug fixes / features)
  - [x] (If applicable) Documentation has been added / updated (for bug fixes / features)
- [x] HISTORY.rst has been updated (with summary of main changes)
  - [x] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added
- [x] The relevant author information has been added to `AUTHORS.rst` and `.zenodo.json`

### What kind of change does this PR introduce?

* Adds `cf-xarray`-based accessor functions for latitude and longitude
* Implements these accessors for many agroclimatic indices that use `lat` and/or `lon`
* Adds the option to send `"north"` or `"south"` for `lat` to `cool_night_index`
* Cleans up a few bad docstrings

### Does this PR introduce a breaking change?

Yes. The call signature for `cool_night_index` has been affected.

### CODE REVIEW

File: xclim/indices/helpers.py [hunk 1]
```diff
@@ -401,3 +403,53 @@ def wind_speed_height_conversion(
...
+    Returns
+    -------
+    xarray.DataArray
+        Longitude coordinate.
+    """
+    try:
+        lat = da.cf["longitude"]
+        return lat
+    except KeyError as err:
+        n_func = stack()[1].function
```
Comment (aulemahal) [inline xclim/indices/helpers.py:450]: If I was to use this `gather_lon` in a script would the error message work?

If we want to restrict the use of these two new functions, I would suggest making them private, ex: `_gather_lon`.

File: xclim/indices/helpers.py [hunk 2]
```diff
@@ -7,6 +7,8 @@
 """
 from __future__ import annotations
 
+from inspect import stack
+
```
Comment (aulemahal) [inline xclim/indices/helpers.py:11]:
```suggestion
import cf_xarray  # noqa
```
To be sure it is imported before accessing  `.cf`.

File: xclim/indices/_agro.py [hunk 3]
```diff
@@ -437,11 +449,12 @@ def cool_night_index(
     Parameters
     ----------
     tasmin : xarray.DataArray
-      Minimum daily temperature.
-    lat: xarray.DataArray, optional
-      Latitude coordinate.
+        Minimum daily temperature.
+    lat : xarray.DataArray or {"north", "south"}, optional
+        Latitude coordinate.
```
Comment (aulemahal) [inline xclim/indices/_agro.py:454]:
```suggestion
        Latitude coordinate as an array, float or string.
```
I know this looks unnecessary, but the Indicator's docstring doesn't preserve the full annotation and only the choices are shown on RTD. With this extra text, I think it's not clearer what can be passed. See : https://xclim--1180.org.readthedocs.build/en/1180/xclim.indicators.atmos.html#xclim.indicators.atmos._temperature.cool_night_index

**Review by aulemahal — APPROVED** [review]
Good with me!

File: xclim/indices/helpers.py (hunk 1 above)
Comment (Zeitsperre) [reply xclim/indices/helpers.py:450]: Good point. Will address.

Comment (Zeitsperre) [reply xclim/indices/helpers.py:451]: Done.

**Comment by Zeitsperre** [comment]
@aulemahal It looks as though one of our upstream dependencies (the newest xarray@main? I can't reproduce a passing test for some reason) treats time-averaged data differently.

The failing test is due to the fact that `latitude_temperature_index` performs a temporal aggregation on the supplied `tas` file, and drops lots of variable metadata that `cf_xarray` relies on (for `X` and `Y` but not `Z`, oddly).

I'm not sure if this is a bug or a feature. Is this worth reporting upstream?

Before:
```
netcdf tas_Amon_CanESM2_rcp85_r1i1p1_200701-200712 {
dimensions:
        time = UNLIMITED ; // (12 currently)
        bnds = 2 ;
        lat = 64 ;
        lon = 128 ;
variables:
        double time(time) ;
                time:_FillValue = NaN ;
                time:bounds = "time_bnds" ;
                time:axis = "T" ;
                time:long_name = "time" ;
                time:standard_name = "time" ;
                time:units = "days since 1850-01-01" ;
                time:calendar = "365_day" ;
        double time_bnds(time, bnds) ;
                time_bnds:_FillValue = NaN ;
                time_bnds:coordinates = "height" ;
        double lat(lat) ;
                lat:_FillValue = NaN ;
                lat:bounds = "lat_bnds" ;
                lat:units = "degrees_north" ;
                lat:axis = "Y" ;
                lat:long_name = "latitude" ;
                lat:standard_name = "latitude" ;
        double lat_bnds(lat, bnds) ;
                lat_bnds:_FillValue = NaN ;
                lat_bnds:coordinates = "height" ;
        double lon(lon) ;
                lon:_FillValue = NaN ;
                lon:bounds = "lon_bnds" ;
                lon:units = "degrees_east" ;
                lon:axis = "X" ;
                lon:long_name = "longitude" ;
                lon:standard_name = "longitude" ;
        double lon_bnds(lon, bnds) ;
                lon_bnds:_FillValue = NaN ;
                lon_bnds:coordinates = "height" ;
        double height ;
                height:_FillValue = NaN ;
                height:units = "m" ;
                height:axis = "Z" ;
                height:positive = "up" ;
                height:long_name = "height" ;
                height:standard_name = "height" ;
        float tas(time, lat, lon) ;
                tas:_FillValue = 1.e+20f ;
                tas:standard_name = "air_temperature" ;
                tas:long_name = "Near-Surface Air Temperature" ;
                tas:units = "K" ;
                tas:original_name = "ST" ;
                tas:cell_methods = "time: mean (interval: 15 minutes)" ;
                tas:cell_measures = "area: areacella" ;
                tas:history = "2011-03-10T05:13:26Z altered by CMOR: Treated scalar dimension: \'height\'. 2011-03-10T05:13:26Z altered by CMOR: replaced missing value flag (1e+38) with standard missing value (1e+20)." ;
                tas:associated_files = "baseURL: http://cmip-pcmdi.llnl.gov/CMIP5/dataLocation gridspecFile: gridspec_atmos_fx_CanESM2_rcp85_r0i0p0.nc areacella: areacella_fx_CanESM2_rcp85_r0i0p0.nc" ;
                tas:coordinates = "height" ;
                tas:missing_value = 1.e+20f ;


```

After:
```
netcdf test_cf_lat_new {
dimensions:
        lat = 64 ;
        lon = 128 ;
        time = 11 ;
variables:
        double lat(lat) ;
                lat:_FillValue = NaN ;
        double lon(lon) ;
                lon:_FillValue = NaN ;
        double height ;
                height:_FillValue = NaN ;
                height:units = "m" ;
                height:axis = "Z" ;
                height:positive = "up" ;
                height:long_name = "height" ;
                height:standard_name = "height" ;
        int64 time(time) ;
                time:units = "days since 2007-01-01 00:00:00.000000" ;
                time:calendar = "noleap" ;
        float tas(time, lat, lon) ;
                tas:_FillValue = NaNf ;
                tas:coordinates = "height" ;
```

**Comment by aulemahal** [comment]
This is weird indeed! I wouldn't expect a resample on time to remove any information about the (untouched!) spatial dimensions...
It may be worth reporting upstream!

I agree with your fixes though. The indicator already manages the "keep_attrs" mechanism, so your changes shouldn't impact it's behaviour.

**Comment by Zeitsperre** [comment]
See: https://github.com/pydata/xarray/issues/7012

**Comment by Zeitsperre** [comment]
@christian-stepanek

If you're interested in verifying that this fix addresses your needs, please let us know. I believe the new behaviour should address the problems you were having. Best!

**Comment by christian-stepanek** [comment]
> @christian-stepanek
> 
> If you're interested in verifying that this fix addresses your needs, please let us know. I believe the new behaviour should address the problems you were having. Best!

Sorry for coming back late. I downloaded the updated master branch as of today and can confirm that cool night index now successfully identifies latitudes by itself from the data set handed to the function. There is no more error message or warning if only tasmin is given as an argument to that function:

```
>>> df_tasmin = xr.open_dataset(filename).data_vars["t2min"]
>>> cool_night_index=xc.indices.cool_night_index(tasmin=df_tasmin)
>>> 
```
Thanks a lot!!!

**Pull request merged**